### PR TITLE
SOL-51 [Integrate] Integrate smart contract function for sending donation to a campaign

### PR DESCRIPTION
**Pull request by kenorenzsun**
## Issue Link
- Backlog link https://framgiaph.backlog.com/view/SPH_SOLIDITY-51

## Target Page | Component | Function
- `/campaigns/{id}`

## Pre-condition
- none

## Steps | Commands
1. Add here

## Expected Output
- Add here

## Definition of Done
- Create a campaign then get its **id** 
- then put it in the `{id}` of `/campaigns/{id}`
- then input a donation and submit
- after successful donation go to profile page and you will see the amount donated
## Test View Points
- [ ] Add here

## Notes
N/A

## Screenshots | Recordings
[Screencast 2023-07-06 17-46-37.webm](https://github.com/framgia/sph-solidity/assets/116168014/43affbfc-7e74-4a3b-83b5-5fee9636cb59)

### General
- [ ] Does the code meet the requirements of the project?
- [ ] Is the code readable and maintainable?
- [ ] Are there any unnecessary or redundant lines of code?
- [ ] Are there any long methods or functions that should be refactored?
- [ ] Are there any performance issues?

### Functionality
- [ ] Are there any logical errors?
- [ ] Are edge cases handled properly?
- [ ] Are there any potential security vulnerabilities?

### Code Style
- [ ] Does the code adhere to the project's coding standards?
- [ ] Are variable and function names descriptive and meaningful?
- [ ] Is there proper use of comments and documentation?
- [ ] Is the code properly formatted? Are indentation, spacing, and formatting consistent?

### Security
- [ ] Is the website secure from common attacks?
- [ ] Are inputs properly validated and sanitized?
- [ ] Are sensitive data and credentials properly protected?

### Testing
- [ ] Are there sufficient test cases?
- [ ] Do the test cases cover all functionality?
- [ ] Are the test cases properly written and documented?

### Additional Comments
- [ ] Are there any other issues or concerns with the code?
- [ ] Are there any suggestions for improvement?
- [ ] Are there any positive aspects of the code that should be highlighted?

### CODE REVIEW

File: fe/src/components/campaign/fund-form.vue [hunk 1]
```diff
@@ -24,18 +24,35 @@
...
         </p>
       </div>
-      <button type="submit" class="btn-gradient-hr">Fund Campaign</button>
+      <BaseButton
+        type="submit"
+        :class="
+          isLoading
+            ? 'bg-disabled h-9 px-4 rounded-[6px] text-white'
+            : 'btn-gradient-hr'
+        "
```
Comment (chevyB) [inline fe/src/components/campaign/fund-form.vue:35]: Can we add this logic of isLoading class to the basebutton component instead? 
Then pass `isLoading` props

File: fe/src/components/campaign/fund-form.vue [hunk 2]
```diff
@@ -58,11 +75,19 @@ const handleChange = (e: any) => {
...
 const { value: amount } = useField<FundCampaignRequest["amount"]>("amount");
 
 const onSubmit = handleSubmit(() => {
-  console.log(values);
-  resetForm();
+  if (isConnected) {
+    emit("fund-campaign", values.amount);
+    resetForm();
+  } else {
+    warning();
```
Comment (chevyB) [inline fe/src/components/campaign/fund-form.vue]: i think no need to create a function for now for `warning()`. Unless it has so many content

File: fe/src/pages/campaigns/[id].vue [hunk 3]
```diff
@@ -25,17 +25,60 @@
...
 
 const campaignValueSample = ref<CampaignProps>(CampaignSample);
 
+const isLoading = ref<boolean>(false);
+
+const useWallet: any = useWalletStore();
+const { isConnected } = storeToRefs(useWallet);
+const { id } = route.params;
+
+const donateCampaign = async (amount: number) => {
```
Comment (chevyB) [inline fe/src/pages/campaigns/[id].vue]:
```suggestion
const donateCampaign = async (amount: number) : void => {
```

**Review by chevyB — CHANGES_REQUESTED** [review]

**Review by wylenjoan — COMMENTED** [review]
Can you make the cards in `My Campaigns` tab of Profile page clickable and redirect to individual Campaign page? 🙇

File: fe/src/components/campaign/fund-form.vue [hunk 4]
```diff
@@ -24,18 +24,33 @@
...
-      <button type="submit" class="btn-gradient-hr">Fund Campaign</button>
+      <BaseButton
+        type="submit"
+        loading-class="bg-disabled h-9 px-4 rounded-[6px] text-white"
+        default-class="btn-gradient-hr"
+        :is-loading="isLoading"
+        :disabled="isLoading"
+      >
+        Fund Campaign
+      </BaseButton>
```
Comment (chevyB) [inline fe/src/components/campaign/fund-form.vue]: The new props is not really used by others.
Maybe you can create a new reusable button that is calling basebutton. Like this:

```
const defaultClasss= 'btn-gradient-hr'
const loadingClass = 'bg-disabled h-9 px-4 rounded-[6px] text-white'
<NewButton>
<BaseButton
        type="submit"
        :class="isLoading ? loadingClass : defaultClass"
        :disabled="isLoading"
      >
<NewButton>

```
Above is only applicable if you think there will be similar button in the future.
Else, revert the changes

**Review by chevyB — DISMISSED** [review]

**Review by wylenjoan — DISMISSED** [review]
Approved :white_check_mark:
Review time: 30 mins
Comment: 1
Iteration: 1
Working:
<details>
<summary> Can send donation to campaign </summary>

image

image

</details>

**Comment by ipilikent** [comment]
Comment
• Problem


• Expected Output
- Fund rejected! - when reject
- Funding campaign in progress! when confirm
- Fund successfully send! - finish sent
- Letters cannot be input

**Comment by ipilikent** [comment]
## Approved PR
- The user can send donation in the campaign
- The validation messages are correct and working 
- Can input in the fund field
- The button is working
- After successful donation go to profile page and you will see the amount donated

## Proof
<details>

Sol- 51 Integrate sending donation

</details>

Test Case: https://docs.google.com/spreadsheets/d/114GrEn-LQQ8saIvjz9zqenuk05qmO4dN/edit#gid=896446223
-----
- Total Test time:  1 hr.
- Total Comments: 1
- Total Iterations: 1

**Review by ipilikent — APPROVED** [review]

**Review by wylenjoan — APPROVED** [review]

**Review by chevyB — APPROVED** [review]

**Pull request merged**